controller/setting_controller.go

```diff
@@ -319,12 +313,35 @@ func (sc *SettingController) syncDangerZoneSettingsForManagedComponents(settingN
 	dangerSettingsRequiringSpecificDataEngineVolumesDetached := []types.SettingName{
 		types.SettingNameV1DataEngine,
 		types.SettingNameV2DataEngine,
+		types.SettingNameGuaranteedInstanceManagerCPU,
+		types.SettingNameV2DataEngineGuaranteedInstanceManagerCPU,
 	}
 
 	if slices.Contains(dangerSettingsRequiringSpecificDataEngineVolumesDetached, settingName) {
-		if err := sc.updateDataEngine(settingName); err != nil {
-			return errors.Wrapf(err, "failed to apply %v setting to Longhorn instance managers when there are attached volumes. "+
-				"It will be eventually applied", settingName)
+		switch settingName {
+		case types.SettingNameV1DataEngine, types.SettingNameV2DataEngine:
+			if err := sc.updateDataEngine(settingName); err != nil {
+				return errors.Wrapf(err, "failed to apply %v setting to Longhorn instance managers when there are attached volumes. "+
+					"It will be eventually applied", settingName)
+			}
+		case types.SettingNameGuaranteedInstanceManagerCPU, types.SettingNameV2DataEngineGuaranteedInstanceManagerCPU:
+			dataEngine := longhorn.DataEngineTypeV1
+			if settingName == types.SettingNameV2DataEngineGuaranteedInstanceManagerCPU {
+				dataEngine = longhorn.DataEngineTypeV2
+			}
+
+			detached, _, err := sc.ds.AreAllVolumesDetached(dataEngine)
+			if err != nil {
+				return errors.Wrapf(err, "failed to check volume detachment for %v setting update", settingName)
+			}
+
+			if !detached {
+				return &types.ErrorInvalidState{Reason: fmt.Sprintf("failed to apply %v setting to Longhorn components when there are attached volumes. It will be eventually applied", settingName)}
+			}
+
+			if err := sc.updateInstanceManagerCPURequest(); err != nil {
```

Comment:
The `updateInstanceManagerCPURequest` will trigger the recreation of all IM pods regardless of whether it's v1 or v2. For example, if the v1 setting is changed but v2 is also enabled, then the v2 IM pods will be restarted as well.

Reply:
No, it will not restart v2 IM pods.
The changing resource request is handled by [GetInstanceManagerCPURequirement](https://github.com/longhorn/longhorn-manager/blob/master/controller/setting_controller.go#L1238C74-L1238C74) in `updateInstanceManagerCPURequest` and it will handle v1 and v2 volumes separately

Reply:
I meant all IM pods will be walked through instead of filtering the data engine type initially.

[/controller/setting_controller.go#L1208-L1233](https://github.com/longhorn/longhorn-manager/blob/98189bfad382f206c0a6ba13e1dbde6283707795/controller/setting_controller.go#L1208-L1233)
```go
func (sc *SettingController) updateInstanceManagerCPURequest() error {
	imPodList, err := sc.ds.ListInstanceManagerPods()
	if err != nil {
		return errors.Wrap(err, "failed to list instance manager pods for toleration update")
	}
	imMap, err := sc.ds.ListInstanceManagersRO()
	if err != nil {
		return err
	}

	for _, imPod := range imPodList {
		if _, exists := imMap[imPod.Name]; !exists {
			continue
		}
		lhNode, err := sc.ds.GetNode(imPod.Spec.NodeName)
		if err != nil {
			return err
		}
		if types.GetCondition(lhNode.Status.Conditions, longhorn.NodeConditionTypeReady).Status != longhorn.ConditionStatusTrue {
			continue
		}

		resourceReq, err := GetInstanceManagerCPURequirement(sc.ds, imPod.Name)
		if err != nil {
			return err
		}
```

Reply:
Modified to pass the `dataEngine` to `updateInstanceManagerCPURequest` and then List the pods and the instance managers by the `dataEgnine`.
Thanks.